### PR TITLE
WIP: Simd optimization documentation

### PR DESCRIPTION
**Pull request by seiko2plus**
Trying to put all ideas and explanation together,
maybe moved part or all of it later to the main NEP doc
or into a separate document.

The current format is MD, since there's no intention
to directly merge it against NumPy main branch.

### CODE REVIEW

**Comment by seiko2plus** [comment]
@mattip, can I proceed?

**Comment by mattip** [comment]
I think this is too much detail for the NEP. It would be good as a follow-on document in the reference documentation, maybe @rgommers has an idea how this would fit into the current documentation. 

It would be great if the tables were generated from the source code so they would always be synced with the actual code.

**Comment by rgommers** [comment]
I would choose to put it as a separate section in the reference guide, at the same level as (and in between) "NumPy internals" and "NumPy and SWIG" in https://numpy.org/devdocs/reference/index.html. There may be a small piece that could fit into the `numpy.distutils` part, but let's keep it all together - that's a lot easier for now.

> It would be great if the tables were generated from the source code so they would always be synced with the actual code.

Only if it's easy, happy to defer that and open an issue that anyone can try and tackle later - it's not a blocker.

All of those topics look interesting, but if you need to prioritize, I'd be most interested in the first sections under `Understanding CPU Dispatching`.

**Comment by seiko2plus** [comment]
@rgommers, I think @mattip is right about it, implied features tables must be auto-generated since there're differences from compiler to another even within the same architecture and I have to explain it in **special cases** section, for example, `FMA3` can't be enabled without `AVX2` in msvc and Intel compilers, so `FMA3` force implies `AVX2` on these compilers. I can easily handle it, by following the same way in the [testing unit](https://github.com/seiko2plus/numpy/blob/de64d35cbbcffeabd0e5f64ec3b4bd74990ca8b1/numpy/distutils/tests/test_ccompiler_opt.py).

File: doc/source/reference/simd-optimizations.rst [hunk 1]
```diff
@@ -111,85 +80,477 @@ The following tables show the current supported optimizations sorted from the lo
...
+
+| **Dispatch-able features** is our dispatched set of additional
+  optimizations that had been configured via
+| ``--cpu-dispatch``. They're not activated by default and always guard
+  it by other C definitions prefixed with
+| ``NPY__CPU_TARGET_``. C definitions ``NPY__CPU_TARGET_`` are only
+  enabled within **dispatch-able sources**.
+
+6- Dispatch-able sources and configuration statements
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (seiko2plus) [inline doc/source/reference/simd-optimizations.rst:302]: @mattip, @rgommers, I wonder if such a thing like that can be understandable?
Any ideas!!

**Comment by mattip** [comment]
Thanks @seiko2plus

**Pull request merged**